nbsite/nbbuild.py

```diff
@@ -178,34 +187,170 @@ class FixNotebookLinks(Preprocessor):
+        """Gets potential link targets corresponding to a notebook file path.
+        Potential link targets (source files paths) are formed by using the
+        relative path of the link target notebook, and adding the file
+        extensions from `cls.file_types`. In addition if the the filename part
+        of the `nb_filepath` starts with digits + one of ('-', '_', ' '),
+        the stem without the number prefix is used to form a potential link
+        target.
+
+        Parameters
+        ----------
+        nb_filepath:
+            The filepath of a link target a notebook. Must end with ".ipynb".
+            The links are relative to the directory of the notebook which
+            contains the markdown link. Example: "../foo/10-Some.ipynb"
+
+        Example
+        -------
+        >>> list(_iter_source_file_candidates("../foo/01-Some_Notebook.ipynb"))
+        [
+            "../foo/01-Some_Notebook.rst",
+            "../foo/01-Some_Notebook.md",
+            "../foo/Some_Notebook.rst",
+            "../foo/Some_Notebook.md",
+        ],
+        """
+
+        if not nb_filepath.endswith('.ipynb'):
+            raise ValueError('nb_filepath must end with .ipynb')
+
+        nb_path_without_extension = nb_filepath[:-6]
+        for extension in cls.file_types:
+            yield f"{nb_path_without_extension}.{extension}"
+
+        for extension in cls.file_types:
+            *directory, stem = re.split('(/)', nb_path_without_extension)
+            match = re.match(r"\d+[ -_](.*)", stem)
+
+            if match:
+                yield ''.join(directory) + match.group(1) + '.' + extension
+
+
+    @staticmethod
+    def _create_target_link(nb_link: str, target_relpath:str) -> str:
+        """Using a markdown link `nb_link`, create a markdown link which
+        changes the link target file to be `target_relpath`.
+
+        Example
+        -------
+        If nb_link is "[sometext](../../somepath/01-subject.ipynb#spam)", and
+        the target_filename is "../otherpath/subject.rst", returns
+        "[sometext](../otherpath/subject.rst#spam)"
+
+        Notes:
+        * If you pass as nb_link text which contains no markdown links, a
+          ValueError will be raised
+        * If you pass as nb_link text which contains more than one markdown
+          links, the behaviour is undefined.
+        """
+
+        # Example: [sometext](../../somepath/subject.ipynb#spam)
+        match = re.match(
+            (
+                r"^(\[.+?\]\()"  # [sometext](
+                r"(.+?)"         #  ../../somepath/subject
+                r"(\.ipynb)"     # .ipynb
+                r"(#*?.*?\))"    # #spam)
+            ),
+            nb_link,
+        )
+        if not match:
+            raise ValueError(f'Not a valid markdown link!: {nb_link}')
+
+        return f"{match.group(1)}{target_relpath}{match.group(4)}"
+
+    @staticmethod
+    def file_exists(file_path: str) -> bool:
+        return os.path.isfile(file_path)
```

Comment:
I can see this is done for mocking in tests, can you add a small comment about that?

Reply:
Yeah sure. This was one of the things which had quite many options

1. A method on the class
2. A function on module level
3. Just use os.path.isfile and mock it in the tests

Could have also done option 3, but was thinking that it makes the test perhaps bit more fragile, if for example the implementation would start using pathlib.Path and its exists method. But probably also that would have been fine.

Added a comment and made the method private.

Reply:
I was about to suggest option 3, but after thinking about it I came to the conclusion that the current implementation is fine in this case.